File: Themer/src/main/java/com/aliucord/plugins/themer/ThemerSettings.java
/*
 * Ven's Aliucord Plugins
 * Copyright (C) 2021 Vendicated
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at 
 * http://www.apache.org/licenses/LICENSE-2.0
*/

package com.aliucord.plugins.themer;


import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Environment;
import android.view.View;
import android.webkit.MimeTypeMap;
import android.widget.TextView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;

import com.aliucord.*;
import com.aliucord.fragments.InputDialog;
import com.aliucord.fragments.SettingsPage;
import com.aliucord.plugins.Themer;
import com.aliucord.views.*;
import com.discord.views.CheckedSetting;
import com.discord.views.RadioManager;
import com.lytefast.flexinput.R;

import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.File;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;

import kotlin.jvm.functions.Function1;

public class ThemerSettings extends SettingsPage {
    private static final String THEME = "theme";
    private static final String FONT = "font";

    private ActivityResultLauncher<Intent> launcher;
    private String intentType;

    private File importComponent(Uri uri, String name, String ext) throws Throwable {
        if (!name.endsWith(ext)) name += ext;
        var file = new File(Constants.BASE_PATH, "themes/" + name);
        try (var is = requireActivity().getContentResolver().openInputStream(uri);
             var os = new FileOutputStream(file)
        ) {
            int n;
            byte[] buf = new byte[16384]; // 16 KB
            while ((n = is.read(buf)) > -1) {
                os.write(buf, 0, n);
            }
            os.flush();
        }
        return file;
    }

    private void handleImport(String ext, String kind) {
        var type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext);
        var intent = new Intent(Intent.ACTION_GET_CONTENT).setType(type);
        intent = Intent.createChooser(intent, "Choose a file");
        intentType = kind;
        launcher.launch(intent);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onViewBound(View view) {
        super.onViewBound(view);

        var ctx = requireContext();

        var act = requireActivity();
        if (launcher == null)
            launcher = act.registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                res -> {
                    if (res.getResultCode() == Activity.RESULT_OK && res.getData() != null) {
                        var data = res.getData();
                        var contentUri = data.getData();

                        var isFont = intentType.equals(FONT);
                        var ext = isFont ? ".ttf" : ".json";

                        Function1<String, Object> doImport = name -> {
                            try {
                                var file = importComponent(contentUri, name, ext);
                                Utils.showToast(ctx, "Successfully imported " + intentType);
                                if (!isFont)         {
                                    ThemeManager.themes.clear();
                                    ThemeManager.loadThemes(ctx, false, false);
                                } else {
                                    ThemeManager.settings.setString("font", file.getAbsolutePath());
                                    ThemeManager.loadFont(-1, file, true);
                                }
                                reRender();
                            } catch (Throwable th) {
                                Themer.logger.error(ctx, "Failed to import " + intentType, th);
                            }
                            return null;
                        };

                        if (contentUri.toString().endsWith(ext)) doImport.invoke(new File(contentUri.getPath()).getName());
                        else {
                            try {
                                var json = new JSONObject(new JSONTokener(new String(Utils.readBytes(act.getContentResolver().openInputStream(contentUri)))));
                                if (json.has("name")) {
                                    doImport.invoke(json.getString("name"));
                                    return;
                                }
                            } catch (Throwable ignored) { }

                            var dialog = new InputDialog()
                                    .setPlaceholderText("Filename")
                                    .setTitle("Filename")
                                    .setDescription("Please specify a name for this " + (isFont ? "Font" : "Theme"));
                            dialog.setOnOkListener(e -> {
                                var text = dialog.getInput();
                                if (text.length() > 0) {
                                    doImport.invoke(text);
                                    dialog.dismiss();
                                }
                            });
                            dialog.show(getParentFragmentManager(), "Themer");
                        }
                    }
                }
        );

        setActionBarTitle("Themer");

        var exportBtn = new ToolbarButton(ctx);
        exportBtn.setOnClickListener(v -> {
            var f = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS), "nameToColor.json");
            try (var fos = new FileOutputStream(f)) {
                var json = new JSONObject();
                var res = v.getResources();
                var theme = v.getContext().getTheme();
                for (var field : R.c.class.getDeclaredFields()) {
                    String colorName = field.getName();
                    int colorId = field.getInt(null);
                    int color = res.getColor(colorId, theme);
                    if (color == 0) continue;
                    json.put(colorName, color);
                }
                fos.write(json.toString(2).getBytes(StandardCharsets.UTF_8));
                Utils.showToast(v.getContext(), "Successfully created nameToColor.json");
            } catch (Throwable ignored) { }
        });
        exportBtn.setImageDrawable(ContextCompat.getDrawable(ctx, R.d.ic_theme_24dp));
        exportBtn.getDrawable().mutate().setAlpha(0);
        addHeaderButton(exportBtn);

        var fontBtn = new Button(ctx);
        fontBtn.setText("Choose font");
        fontBtn.setOnClickListener(e -> handleImport("ttf", FONT));

        var importBtn = new Button(ctx);
        importBtn.setText("Import theme");
        importBtn.setOnClickListener(e -> handleImport("json", THEME));

        var refreshBtn = new Button(ctx);
        refreshBtn.setText("Load missing themes");
        refreshBtn.setOnClickListener(e -> {
            ThemeManager.themes.clear();
            ThemeManager.loadThemes(e.getContext(), false, false);
            reRender();
        });

        addView(importBtn);
        addView(refreshBtn);
        addView(new Divider(ctx));

        var header = new TextView(ctx, null, 0, R.h.UiKit_Settings_Item_Header);
        header.setText("Active Theme");
        header.setTypeface(ResourcesCompat.getFont(ctx, Constants.Fonts.whitney_semibold));

        addView(header);

        if (ThemeManager.themes.size() == 0) {
            var text = new TextView(ctx, null, 0, R.h.UiKit_TextView);
            text.setText("Hmm... No themes found.");
            addView(text);
        } else {
            var transparencySwitch = Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.SWITCH, "Transparency", "Enables transparency for themes with custom background. Requires restart");
            transparencySwitch.setChecked(ThemeManager.settings.getBool("enableTransparency", false));
            transparencySwitch.setOnCheckedListener(c -> ThemeManager.settings.setBool("enableTransparency", c));
            addView(transparencySwitch);

            var items = CollectionUtils.map(ThemeManager.themes, t -> Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.RADIO, String.format("%s v%s by %s", t.name, t.version, t.author), null));
            var noTheme = Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.RADIO, "None", null);

            items.add(0, noTheme);

            var manager = new RadioManager(items);
            manager.a(noTheme);
            noTheme.e(e -> {
                ThemeManager.activeTheme = null;
                int idx = manager.b();
                if (idx != 0) {
                    ThemeManager.themes.get(idx - 1).disable();
                    manager.a(noTheme);
                    Utils.appActivity.recreate();
                    requireActivity().onBackPressed();
                    Utils.openPageWithProxy(ctx, new ThemerSettings());
                }
            });

            addView(noTheme);

            for (int i = 1; i < items.size(); i++) {
                var theme = ThemeManager.themes.get(i - 1);
                var item = items.get(i);
                final int idx = i;
                item.e(e -> {
                    if (manager.b() != idx) {
                        if (theme.load()) {
                            theme.enable();
                            manager.a(item);
                            requireActivity().onBackPressed();
                            Utils.openPageWithProxy(ctx, new ThemerSettings());
                        } else {
                            Utils.showToast(ctx, "Something went wrong while loading that theme :(");
                        }
                    }
                });
                addView(item);
                if (theme.isEnabled()) manager.a(item);
            }
        }
    }
}

File: Themer/src/main/java/com/aliucord/plugins/themer/ThemerSettings.kt
package com.aliucord.plugins.themer

import android.annotation.SuppressLint
import android.content.Intent
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.view.View
import android.widget.TextView
import androidx.core.content.res.ResourcesCompat
import com.aliucord.*
import com.aliucord.fragments.SettingsPage
import com.aliucord.plugins.Themer
import com.aliucord.utils.ChangelogUtils
import com.aliucord.utils.DimenUtils
import com.aliucord.views.Button
import com.aliucord.views.Divider
import com.discord.views.CheckedSetting
import com.discord.views.RadioManager
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.Snackbar.LENGTH_INDEFINITE
import com.lytefast.flexinput.R
import kotlin.system.exitProcess

class ThemerSettings : SettingsPage() {
    @SuppressLint("SetTextI18n")
    override fun onViewBound(view: View) {
        super.onViewBound(view)

        val ctx = view.context

        setActionBarTitle("Themer")

        TextView(ctx, null, 0, R.h.UiKit_TextView).run {
            val content = "Read the changelog!"
            SpannableStringBuilder(content).let {
                it.setSpan(object : ClickableSpan() {
                    override fun onClick(widget: View) {
                        val manifest = PluginManager.plugins["Themer"]!!.manifest
                        ChangelogUtils.show(context, manifest.version, manifest.changelogMedia, manifest.changelog)
                    }
                }, content.indexOf("changelog"), content.length, SpannableString.SPAN_EXCLUSIVE_EXCLUSIVE)
                text = it
            }
            textAlignment = View.TEXT_ALIGNMENT_CENTER
            DimenUtils.getDefaultPadding().let {
                setPadding(it, it, it, it)
            }
            movementMethod = LinkMovementMethod.getInstance()
            linearLayout.addView(this)
        }

        Button(ctx).apply {
            text = "Load missing themes"
            setOnClickListener {
                ThemeLoader.loadThemes(false)
                reRender()
            }

            linearLayout.addView(this)
        }

        TextView(ctx, null, 0, R.h.UiKit_Settings_Item_Header).apply {
            text = "Transparency Mode"
            typeface = ResourcesCompat.getFont(ctx, Constants.Fonts.whitney_semibold)

            linearLayout.addView(this)
        }

        arrayListOf(
            Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.RADIO, "None", "No transparency"),
            Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.RADIO, "Chat", "Chat is transparent"),
            Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.RADIO, "Chat & Settings", "Chat and Settings page are transparent"),
            Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.RADIO, "Full", "Everything is transparent. NOT IMPLEMENTED YET"),
        ).let { radios ->
            val manager = RadioManager(radios)
            manager.a(radios[Themer.mSettings.transparencyMode.value])
            for (i in 0 until radios.size) {
                val radio = radios[i]
                radio.e {
                    manager.a(radio)
                    Themer.mSettings.transparencyMode = TransparencyMode.from(i)
                    promptRestart(view)
                }
                addView(radio)
            }
        }

        addView(Divider(ctx))

        TextView(ctx, null, 0, R.h.UiKit_Settings_Item_Header).apply {
            text = "Active Theme"
            typeface = ResourcesCompat.getFont(ctx, Constants.Fonts.whitney_semibold)

            linearLayout.addView(this)
        }

        if (ThemeLoader.themes.isEmpty()) {
            TextView(ctx, null, 0, R.h.UiKit_TextView).apply {
                text = "Hmm... No themes found."

                linearLayout.addView(this)
            }
        } else {
            val items = ThemeLoader.themes.map {
                Utils.createCheckedSetting(
                    ctx,
                    CheckedSetting.ViewType.RADIO,
                    "${it.name} v${it.version}",
                    it.author
                )
            }
            val noTheme = Utils.createCheckedSetting(ctx, CheckedSetting.ViewType.RADIO, "None", null)
            (items as MutableList).add(0, noTheme)

            val manager = RadioManager(items)
            manager.a(noTheme)
            noTheme.e {
                ResourceManager.clean()
                val idx = manager.b()
                if (idx != 0) {
                    ThemeLoader.themes[idx - 1].disable()
                    manager.a(noTheme)
                    Utils.appActivity.recreate()
                    promptRestart(view)
                }
            }

            addView(noTheme)

            for (i in 1 until items.size) {
                val theme = ThemeLoader.themes[i - 1]
                val item = items[i]
                item.e {
                    if (manager.b() != i) {
                        if (theme.load()) {
                            theme.enable()
                            manager.a(item)
                            promptRestart(view)
                        } else {
                            Utils.showToast(ctx, "Something went wrong while loading that theme :(")
                        }
                    }
                }
                addView(item)
                if (theme.isEnabled) manager.a(item)
            }
        }
    }

    private fun promptRestart(v: View) {
        Snackbar.make(v, "Changes detected. Restart?", LENGTH_INDEFINITE)
            .setAction("Restart") {
                val ctx = it.context
                ctx.packageManager.getLaunchIntentForPackage(ctx.packageName)?.run {
                    startActivity(Intent.makeRestartActivityTask(component))
                    exitProcess(0)
                }
            }.show()
    }
}
